# Сингулярное разложение матриц

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

In [2]:
A = np.array([[1, 2, 0], [0, 0, 5], [3, -4, 2], [1, 6, 5], [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [3]:
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]


In [4]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]


In [5]:
# Убедимся, что она действительно ортогональна
print(np.dot(U.T, U))

[[ 1. -0.  0.  0.  0.]
 [-0.  1. -0. -0. -0.]
 [ 0. -0.  1.  0. -0.]
 [ 0. -0.  0.  1. -0.]
 [ 0. -0. -0. -0.  1.]]


In [6]:
print(f'Матрица V:\n{V}')

Матрица V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [7]:
# Убедимся, что она действительно ортогональна
print(np.dot(V.T, V))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [8]:
# Проведем проверку
print(np.dot(np.dot(U, D), V.T))

[[ 1.  2. -0.]
 [-0.  0.  5.]
 [ 3. -4.  2.]
 [ 1.  6.  5.]
 [ 0.  1.  0.]]


2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

__Решение__

__а)__ Евклидова норма определяется как 
$$\left \| A \right \|_{E}=\text{max}\left (\frac{\left \| Ax \right \|}{\left \| x \right \|}\right )$$

$$\left \| A \right \|_{E}=\text{max}(1+2+0, 0+0+5, 3+(-4)+2, 1+6+5, 0+1+0)= \text{max}(3, 5, 1, 12, 1)=12$$

__б)__ _Норма Фробениуса_, определяемая как:

$$\left \| A \right \|_{F}=\sqrt{\sum_{i=1}^{m}\sum_{j=1}^{n}a_{ij}^{2}}.$$

$$\left \| A \right \|_{F}=\sqrt{1^2+2^2+5^2+3^2+(-4)^2+2^2+1^2+6^2+5^2+1^2}=\sqrt{122}=11.05$$

или если известно сингулярное разложение матрицы:
$$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$
$$\left \| A \right \|_{F}=\sqrt{8.82^2+6.14^2+2.53^2}=\sqrt{121.8929}=11.04$$